test_rclcpp/test/test_subscription.cpp

```diff
@@ -140,3 +140,50 @@ TEST(CLASSNAME(test_subscription, RMW_IMPLEMENTATION), subscription_and_spinning
   executor.spin_node_some(node);
   ASSERT_EQ(5, counter);
 }
+
+// Shortened version of the test for the ConstSharedPtr callback signature
+TEST(CLASSNAME(test_subscription, RMW_IMPLEMENTATION), subscription_shared_ptr_const) {
+  rclcpp::init(0, nullptr);
+
+  auto node = rclcpp::Node::make_shared("test_subscription");
+
+  rmw_qos_profile_t custom_qos_profile = rmw_qos_profile_default;
+  custom_qos_profile.depth = 10;
+
+  auto publisher = node->create_publisher<test_rclcpp::msg::UInt32>(
+    "test_subscription", custom_qos_profile);
+
+  unsigned long counter = 0;
+  auto callback =
+    [&counter](test_rclcpp::msg::UInt32::ConstSharedPtr msg) -> void
+    {
+      ++counter;
+      printf("  callback() %lu with message data %u\n", counter, msg->data);
+      ASSERT_EQ(counter, msg->data);
+    };
+
+  auto msg = std::make_shared<test_rclcpp::msg::UInt32>();
+  msg->data = 0;
+  rclcpp::executors::SingleThreadedExecutor executor;
+
+  {
+    auto subscriber = node->create_subscription<test_rclcpp::msg::UInt32>(
+      "test_subscription", custom_qos_profile, callback);
+
+    // start condition
+    ASSERT_EQ(0, counter);
+
+    // nothing should be pending here
+    executor.spin_node_some(node);
+    ASSERT_EQ(0, counter);
+
+    msg->data = 1;
+    publisher->publish(msg);
+    ASSERT_EQ(0, counter);
+
+    // wait for the first callback
+    printf("spin_node_once() - callback (1) expected\n");
+    executor.spin_node_once(node);
```

Comment:
This could be flaky in the future. Relying on the next callback being the one you registered above is fragile. I'd wrap this in for loop to try multiple times or maybe use `spin_some`, but that might have its own issues.